classes/submission/form/PKPSubmissionSubmitStep1Form.inc.php

```diff
@@ -78,31 +84,57 @@ function fetch($request) {
 			$templateMgr->assign('copyrightNoticeAgree', true);
 		}
 
-		// Get list of user's author user groups.  If its more than one, we'll need to display an author user group selector
 		$userGroupAssignmentDao = DAORegistry::getDAO('UserGroupAssignmentDAO');
 		$userGroupDao = DAORegistry::getDAO('UserGroupDAO');
-		$authorUserGroupAssignments = $userGroupAssignmentDao->getByUserId($user->getId(), $this->context->getId(), ROLE_ID_AUTHOR);
 		$userGroupNames = array();
-		if (!$authorUserGroupAssignments->wasEmpty()) {
-			while($authorUserGroupAssignment = $authorUserGroupAssignments->next()) {
-				$authorUserGroup = $userGroupDao->getById($authorUserGroupAssignment->getUserGroupId());
-				if ($userGroupDao->userGroupAssignedToStage($authorUserGroup->getId(), WORKFLOW_STAGE_ID_SUBMISSION)) {
-					$userGroupNames[$authorUserGroup->getId()] = $authorUserGroup->getLocalizedName();
-				}
+
+		// List existing user roles
+		$managerUserGroupAssignments = $userGroupAssignmentDao->getByUserId($user->getId(), $this->context->getId(), ROLE_ID_MANAGER);
+		$authorUserGroupAssignments = $userGroupAssignmentDao->getByUserId($user->getId(), $this->context->getId(), ROLE_ID_AUTHOR);
+
+		// List available author roles
+		$availableAuthorUserGroups = $userGroupDao->getUserGroupsByStage($this->context->getId(), WORKFLOW_STAGE_ID_SUBMISSION, ROLE_ID_AUTHOR);
+		while($authorUserGroup = $availableAuthorUserGroups->next()) {
+			if ($authorUserGroup->getPermitSelfRegistration()){
+				$availableUserGroupNames[$authorUserGroup->getId()] = $authorUserGroup->getLocalizedName();
 			}
-			$templateMgr->assign('authorUserGroupOptions', $userGroupNames);
-		} else {
-			// The user doesn't have any author user group assignments.  They should be either a manager.
-			// Add all manager user groups
-			$managerUserGroupAssignments = $userGroupAssignmentDao->getByUserId($user->getId(), $this->context->getId(), ROLE_ID_MANAGER);
-			if($managerUserGroupAssignments) while($managerUserGroupAssignment = $managerUserGroupAssignments->next()) {
+		}
+
+		// Set default group to default author group
+		$defaultGroup = $userGroupDao->getDefaultByRoleId($this->context->getId(), ROLE_ID_AUTHOR);
+		$noExistingRoles = false;
+		$managerGroups = false;
+
+		// If the user has manager roles, add manager roles and available author roles to selection
+		if (!$managerUserGroupAssignments->wasEmpty()) {
+			while($managerUserGroupAssignment = $managerUserGroupAssignments->next()) {
 				$managerUserGroup = $userGroupDao->getById($managerUserGroupAssignment->getUserGroupId());
 				$userGroupNames[$managerUserGroup->getId()] = $managerUserGroup->getLocalizedName();
 			}
+			$managerGroups = join(__('common.listSeparator'), $userGroupNames);
+			$userGroupNames = array_replace($userGroupNames, $availableUserGroupNames);
```

Comment:
> Just theoretically it could be that the $availableUserGroupNames does not exist here. Thus, maybe to declare it as empty array() above?

Thanks, done.

> array_replace means that a manager user group could have the same name as an author user group and in that case we would like only to consider the author user group, right -- we do not want to consider both?

I think that would only happen in the case of them having the same ID, since the array key is the `UserGroup` ID. `array_replace` will preserve keys, whereas `array_merge` will update keys. I'm guessing that's why this @ajnyga used that function.

Reply:
Ah yes, the keys are user group IDs... I think we can leave it as is, although it should not happen that there is the same user group ID in those both arrays...